pkg/cable/libreswan/libreswan.go

```diff
@@ -337,6 +334,15 @@ func whack(args ...string) error {
 // ConnectToEndpoint establishes a connection to the given endpoint and returns a string
 // representation of the IP address of the target endpoint.
 func (i *libreswan) ConnectToEndpoint(endpointInfo *natdiscovery.NATEndpointInfo) (string, error) {
+	if !i.plutoStarted {
+		// Ensure Pluto is started
+		if err := i.runPluto(); err != nil {
+			return "", errors.Wrap(err, "error starting Pluto")
```

Comment:
This is going to change the behaviour. In the current code if `runPluto` returns error, we treat that as a FATAL error as part of Init() method. However, in the updated code its treated as an normal error while connecting to an endpoint. Is this intentional?

Reply:
Good point, I assume that if Pluto fails we should consider it as FATAL.

It looks like  we have the following options:
 
1.  update only Pluto control socket status check , in my Kind environment downtime was ~1.1 Sec with this fix.
2. Seperate Engine '**StartEngine**' API into two functions, to something like **InitEngine** and **StartEngineOperation** ,  **StartEngineOperation** should be called right before  Endpoints handler is started and we should treat failure here as fatal.  for libreswan Pluto should be started at **StartEngineOperation** phase.
3. hack the current code to treat Pluto error as fatal.


I think that since we expect that with **--dpdaction=hold** fix the number of gw pod restarts will decrease significantly, maybe we should go first with '1'

WDYT ?

Reply:
This issue seems to be very specific to libreswan, so Option-2 may not be a good idea IMO. 

That means if we go with Option-1, the downtime will be ~1.1 seconds instead of 0.5 secs? 
Did you get a chance to verify this behaviour on an OCP deployment?

Ideally when downtime cannot be avoided due to some limitations in Libreswan driver, the lower the duration the better as applications can sometimes tolerate that downtime. Can you please try once on an OCP deployment and let us know your findings?

Also, I'd like to hear from @skitt if he has any feedback on this.

Reply:
I agree we shouldn’t change the API for this. Looking at this purely from the perspective of the Libreswan cable driver, not being able to start Pluto is a fatal error, but the behaviour of the agent should be decided based on what the remediation should be.

It seems to me that the safest option is to stick to the current behaviour, and abort the gateway pod if Pluto can’t be started (moving this to a just-in-time start doesn’t change much).

If we want to reconsider things:

* if we abort the gateway pod, we’ll retry starting Pluto later on, after some delay (while the pod restart)
* if we return an error, we’ll retry starting Pluto too, the next time the connections are synced (5 sec?)

So crashing the pod might actually result in a quicker retry, and possibly fix issues preventing Pluto from starting. Put another way, it seems to me that restarting later on from the same pod is unlikely to actually find different conditions, whereas restarting the entire pod could fix things.

In summary, I vote for keeping the current behaviour: move the code as suggested in the PR, but make failure to run Pluto fatal as it was before.

Reply:
Delaying this PR  till we get  feedback on https://bugzilla.redhat.com/show_bug.cgi?id=2216154

Reply:
So, following the discussion on ^^ BZ, we identified two issues related to Libreswan datapath downtime when GW pod was restarted.

1. Datapath is down from the moment Pluto is started until endpoints are reconfigured  
2. When Libreswan is configured to use UDP encapsulation, traffic is down as long as Pluto not running.


This PR handles '1' by running Pluto just before configuring the endpoints.

cc @skitt @sridhargaddam